NPE in CompassActivity

Similar to #7303 or #7300 but occuring in Compass Activity:

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:4290)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:4330)
  at android.app.servertransaction.ResumeActivityItem.execute (ResumeActivityItem.java:51)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:145)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:70)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2123)
  at android.os.Handler.dispatchMessage (Handler.java:109)
  at android.os.Looper.loop (Looper.java:207)
  at android.app.ActivityThread.main (ActivityThread.java:7470)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:524)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:958)
Caused by: java.lang.NullPointerException: 
  at io.reactivex.internal.functions.ObjectHelper.requireNonNull (ObjectHelper.java:39)
  at io.reactivex.Observable.combineLatest (Observable.java:433)
  at cgeo.geocaching.sensors.GeoDirHandler.start (GeoDirHandler.java:111)
  at cgeo.geocaching.sensors.GeoDirHandler.start (GeoDirHandler.java:77)
  at cgeo.geocaching.CompassActivity.onResume (CompassActivity.java:128)
  at android.app.Instrumentation.callActivityOnResume (Instrumentation.java:1456)
  at android.app.Activity.performResume (Activity.java:7592)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:4272)
ˋˋˋ

java.lang.RuntimeException: 

at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3790)

at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:3830)

at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:3038)

at android.app.ActivityThread.-wrap11 (Unknown Source)

at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1696)

at android.os.Handler.dispatchMessage (Handler.java:105)

at android.os.Looper.loop (Looper.java:164)

at android.app.ActivityThread.main (ActivityThread.java:6944)

at java.lang.reflect.Method.invoke (Native Method)

at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:327)

at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1374)

Caused by: java.lang.NullPointerException: 

at io.reactivex.internal.functions.ObjectHelper.requireNonNull (ObjectHelper.java:39)

at io.reactivex.Observable.combineLatest (Observable.java:433)

at cgeo.geocaching.sensors.GeoDirHandler.start (GeoDirHandler.java:111)

at cgeo.geocaching.sensors.GeoDirHandler.start (GeoDirHandler.java:77)

at cgeo.geocaching.CompassActivity.onResume (CompassActivity.java:128)

at android.app.Instrumentation.callActivityOnResume (Instrumentation.java:1361)

at android.app.Activity.performResume (Activity.java:7361)

at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3765)

--- a/main/src/cgeo/geocaching/CacheDetailActivity.java
+++ b/main/src/cgeo/geocaching/CacheDetailActivity.java
@@ -44,13 +44,12 @@ import cgeo.geocaching.network.AndroidBeam;
 import cgeo.geocaching.network.HtmlImage;
 import cgeo.geocaching.network.Network;
 import cgeo.geocaching.network.SmileyImage;
-import cgeo.geocaching.permission.PermissionGrantedCallback;
 import cgeo.geocaching.permission.PermissionHandler;
 import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.playservices.AppInvite;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
-import cgeo.geocaching.sensors.Sensors;
 import cgeo.geocaching.settings.Settings;
 import cgeo.geocaching.staticmaps.StaticMapsProvider;
 import cgeo.geocaching.storage.DataStore;
@@ -417,15 +416,10 @@ public class CacheDetailActivity extends AbstractViewPagerActivity<CacheDetailAc
         super.onResume();
 
         // resume location access
-        PermissionHandler.executeIfLocationPermissionGranted(this, new PermissionGrantedCallback(PermissionRequestContext.CacheDetailActivity) {
+        PermissionHandler.executeIfLocationPermissionGranted(this, new RestartLocationPermissionGrantedCallback(PermissionRequestContext.CacheDetailActivity) {
 
             @Override
-            public void execute() {
-                Log.d("CacheDetailActivity.OnResumePermissionGrantedCallback.execute");
-                final Sensors sensors = Sensors.getInstance();
-                sensors.setupGeoDataObservables(Settings.useGooglePlayServices(), Settings.useLowPowerMode());
-                sensors.setupDirectionObservable();
-
+            public void executeAfter() {
                 startOrStopGeoDataListener(true);
             }
         });

--- a/main/src/cgeo/geocaching/CacheListActivity.java
+++ b/main/src/cgeo/geocaching/CacheListActivity.java
@@ -53,9 +53,9 @@ import cgeo.geocaching.network.Cookies;
 import cgeo.geocaching.network.DownloadProgress;
 import cgeo.geocaching.network.Network;
 import cgeo.geocaching.network.Send2CgeoDownloader;
-import cgeo.geocaching.permission.PermissionGrantedCallback;
 import cgeo.geocaching.permission.PermissionHandler;
 import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
 import cgeo.geocaching.sensors.Sensors;
@@ -629,15 +629,10 @@ public class CacheListActivity extends AbstractListActivity implements FilteredA
         super.onResume();
 
         // resume location access
-        PermissionHandler.executeIfLocationPermissionGranted(this, new PermissionGrantedCallback(PermissionRequestContext.CacheListActivity) {
+        PermissionHandler.executeIfLocationPermissionGranted(this, new RestartLocationPermissionGrantedCallback(PermissionRequestContext.CacheListActivity) {
 
             @Override
-            public void execute() {
-                Log.d("CacheDetailActivity.OnResumePermissionGrantedCallback.execute");
-                final Sensors sensors = Sensors.getInstance();
-                sensors.setupGeoDataObservables(Settings.useGooglePlayServices(), Settings.useLowPowerMode());
-                sensors.setupDirectionObservable();
-
+            public void executeAfter() {
                 resumeDisposables.add(geoDirHandler.start(GeoDirHandler.UPDATE_GEODATA | GeoDirHandler.UPDATE_DIRECTION | GeoDirHandler.LOW_POWER, 250, TimeUnit.MILLISECONDS));
             }
         });

--- a/main/src/cgeo/geocaching/CompassActivity.java
+++ b/main/src/cgeo/geocaching/CompassActivity.java
@@ -8,6 +8,9 @@ import cgeo.geocaching.log.LoggingUI;
 import cgeo.geocaching.maps.DefaultMap;
 import cgeo.geocaching.models.Geocache;
 import cgeo.geocaching.models.Waypoint;
+import cgeo.geocaching.permission.PermissionHandler;
+import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
 import cgeo.geocaching.sensors.GpsStatusProvider.Status;
@@ -125,8 +128,20 @@ public class CompassActivity extends AbstractActionBarActivity {
 
     @Override
     public void onResume() {
-        onResume(geoDirHandler.start(GeoDirHandler.UPDATE_GEODIR),
-                Sensors.getInstance().gpsStatusObservable().observeOn(AndroidSchedulers.mainThread()).subscribe(gpsStatusHandler));
+        super.onResume();
+
+        // resume location access
+        PermissionHandler.executeIfLocationPermissionGranted(this,
+                new RestartLocationPermissionGrantedCallback(PermissionRequestContext.CompassActivity) {
+
+                    @Override
+                    public void executeAfter() {
+                        resumeDisposables(geoDirHandler.start(GeoDirHandler.UPDATE_GEODIR),
+                                Sensors.getInstance().gpsStatusObservable().observeOn(AndroidSchedulers.mainThread()).subscribe(gpsStatusHandler));
+                    }
+                });
+
+
         forceRefresh();
     }
 

--- a/main/src/cgeo/geocaching/EditWaypointActivity.java
+++ b/main/src/cgeo/geocaching/EditWaypointActivity.java
@@ -14,6 +14,9 @@ import cgeo.geocaching.models.CalcState;
 import cgeo.geocaching.models.Geocache;
 import cgeo.geocaching.models.Waypoint;
 import cgeo.geocaching.network.SmileyImage;
+import cgeo.geocaching.permission.PermissionHandler;
+import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
 import cgeo.geocaching.settings.Settings;
@@ -286,7 +289,17 @@ public class EditWaypointActivity extends AbstractActionBarActivity implements C
 
     @Override
     public void onResume() {
-        super.onResume(geoDirHandler.start(GeoDirHandler.UPDATE_GEODATA));
+        super.onResume();
+
+        // resume location access
+        PermissionHandler.executeIfLocationPermissionGranted(this,
+                new RestartLocationPermissionGrantedCallback(PermissionRequestContext.EditWaypointActivity) {
+
+                    @Override
+                    public void executeAfter() {
+                        resumeDisposables(geoDirHandler.start(GeoDirHandler.UPDATE_GEODATA));
+                    }
+                });
     }
 
     @Override

--- a/main/src/cgeo/geocaching/NavigateAnyPointActivity.java
+++ b/main/src/cgeo/geocaching/NavigateAnyPointActivity.java
@@ -7,6 +7,9 @@ import cgeo.geocaching.location.DistanceParser;
 import cgeo.geocaching.location.Geopoint;
 import cgeo.geocaching.location.GeopointFormatter;
 import cgeo.geocaching.models.Destination;
+import cgeo.geocaching.permission.PermissionHandler;
+import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
 import cgeo.geocaching.sensors.Sensors;
@@ -239,7 +242,18 @@ public class NavigateAnyPointActivity extends AbstractActionBarActivity implemen
 
     @Override
     public void onResume() {
-        super.onResume(geoDirHandler.start(GeoDirHandler.UPDATE_GEODATA));
+        super.onResume();
+
+        // resume location access
+        PermissionHandler.executeIfLocationPermissionGranted(this,
+                new RestartLocationPermissionGrantedCallback(PermissionRequestContext.NavigateAnyPointActivity) {
+
+                    @Override
+                    public void executeAfter() {
+                        resumeDisposables(geoDirHandler.start(GeoDirHandler.UPDATE_GEODATA));
+                    }
+                });
+
         init();
     }
 

--- a/main/src/cgeo/geocaching/TrackableActivity.java
+++ b/main/src/cgeo/geocaching/TrackableActivity.java
@@ -13,12 +13,11 @@ import cgeo.geocaching.log.TrackableLogsViewCreator;
 import cgeo.geocaching.models.Trackable;
 import cgeo.geocaching.network.AndroidBeam;
 import cgeo.geocaching.network.HtmlImage;
-import cgeo.geocaching.permission.PermissionGrantedCallback;
 import cgeo.geocaching.permission.PermissionHandler;
 import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
-import cgeo.geocaching.sensors.Sensors;
 import cgeo.geocaching.settings.Settings;
 import cgeo.geocaching.ui.AbstractCachingPageViewCreator;
 import cgeo.geocaching.ui.AnchorAwareLinkMovementMethod;
@@ -207,15 +206,11 @@ public class TrackableActivity extends AbstractViewPagerActivity<TrackableActivi
         super.onResume();
 
         // resume location access
-        PermissionHandler.executeIfLocationPermissionGranted(this, new PermissionGrantedCallback(PermissionRequestContext.TrackableActivity) {
+        PermissionHandler.executeIfLocationPermissionGranted(this,
+                new RestartLocationPermissionGrantedCallback(PermissionRequestContext.TrackableActivity) {
 
             @Override
-            public void execute() {
-                Log.d("TrackableActivity.OnResumePermissionGrantedCallback.execute");
-                final Sensors sensors = Sensors.getInstance();
-                sensors.setupGeoDataObservables(Settings.useGooglePlayServices(), Settings.useLowPowerMode());
-                sensors.setupDirectionObservable();
-
+            public void executeAfter() {
                 if (!Settings.useLowPowerMode()) {
                     geoDataDisposable.add(locationUpdater.start(GeoDirHandler.UPDATE_GEODATA));
                 }

--- a/main/src/cgeo/geocaching/activity/AbstractActivity.java
+++ b/main/src/cgeo/geocaching/activity/AbstractActivity.java
@@ -79,8 +79,7 @@ public abstract class AbstractActivity extends ActionBarActivity implements IAbs
         return super.onOptionsItemSelected(item);
     }
 
-    public void onResume(final Disposable... resumeDisposable) {
-        super.onResume();
+    protected void resumeDisposables(final Disposable... resumeDisposable) {
         this.resumeDisposable.addAll(resumeDisposable);
     }
 

--- a/main/src/cgeo/geocaching/maps/CGeoMap.java
+++ b/main/src/cgeo/geocaching/maps/CGeoMap.java
@@ -31,9 +31,9 @@ import cgeo.geocaching.maps.routing.RoutingMode;
 import cgeo.geocaching.models.Geocache;
 import cgeo.geocaching.models.Waypoint;
 import cgeo.geocaching.network.AndroidBeam;
-import cgeo.geocaching.permission.PermissionGrantedCallback;
 import cgeo.geocaching.permission.PermissionHandler;
 import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
 import cgeo.geocaching.sensors.Sensors;
@@ -530,18 +530,14 @@ public class CGeoMap extends AbstractMap implements ViewFactory {
         super.onResume();
 
         // resume location access
-        PermissionHandler.executeIfLocationPermissionGranted(this.activity, new PermissionGrantedCallback(PermissionRequestContext.CacheDetailActivity) {
+        PermissionHandler.executeIfLocationPermissionGranted(this.activity,
+                new RestartLocationPermissionGrantedCallback(PermissionRequestContext.CGeoMap) {
 
-            @Override
-            public void execute() {
-                Log.d("CGeoMap.OnResumePermissionGrantedCallback.execute");
-                final Sensors sensors = Sensors.getInstance();
-                sensors.setupGeoDataObservables(Settings.useGooglePlayServices(), Settings.useLowPowerMode());
-                sensors.setupDirectionObservable();
-
-                resumeDisposables.addAll(geoDirUpdate.start(GeoDirHandler.UPDATE_GEODIR), startTimer());
-            }
-        });
+                    @Override
+                    public void executeAfter() {
+                        resumeDisposables.addAll(geoDirUpdate.start(GeoDirHandler.UPDATE_GEODIR), startTimer());
+                    }
+                });
 
         final List<String> toRefresh;
         synchronized (dirtyCaches) {

--- a/main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java
+++ b/main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java
@@ -38,9 +38,9 @@ import cgeo.geocaching.maps.mapsforge.v6.layers.TapHandlerLayer;
 import cgeo.geocaching.maps.routing.Routing;
 import cgeo.geocaching.maps.routing.RoutingMode;
 import cgeo.geocaching.models.Geocache;
-import cgeo.geocaching.permission.PermissionGrantedCallback;
 import cgeo.geocaching.permission.PermissionHandler;
 import cgeo.geocaching.permission.PermissionRequestContext;
+import cgeo.geocaching.permission.RestartLocationPermissionGrantedCallback;
 import cgeo.geocaching.sensors.GeoData;
 import cgeo.geocaching.sensors.GeoDirHandler;
 import cgeo.geocaching.sensors.Sensors;
@@ -785,18 +785,14 @@ public class NewMap extends AbstractActionBarActivity implements XmlRenderThemeM
         }
 
         // resume location access
-        PermissionHandler.executeIfLocationPermissionGranted(this, new PermissionGrantedCallback(PermissionRequestContext.CacheDetailActivity) {
+        PermissionHandler.executeIfLocationPermissionGranted(this,
+                new RestartLocationPermissionGrantedCallback(PermissionRequestContext.NewMap) {
 
-            @Override
-            public void execute() {
-                Log.d("NewMap.initializeLayersPermissionGrantedCallback.execute");
-                final Sensors sensors = Sensors.getInstance();
-                sensors.setupGeoDataObservables(Settings.useGooglePlayServices(), Settings.useLowPowerMode());
-                sensors.setupDirectionObservable();
-
-                resumeDisposables.add(geoDirUpdate.start(GeoDirHandler.UPDATE_GEODIR));
-            }
-        });
+                    @Override
+                    public void executeAfter() {
+                        resumeDisposables.add(geoDirUpdate.start(GeoDirHandler.UPDATE_GEODIR));
+                    }
+                });
     }
 
     @Override

--- a/main/src/cgeo/geocaching/permission/PermissionRequestContext.java
+++ b/main/src/cgeo/geocaching/permission/PermissionRequestContext.java
@@ -9,7 +9,12 @@ public enum PermissionRequestContext {
     MainActivityStorage(1113, R.string.storage_permission_request_explanation),
     TrackableActivity(2222, R.string.location_permission_request_explanation),
     CacheDetailActivity(3333, R.string.location_permission_request_explanation),
-    CacheListActivity(4444, R.string.location_permission_request_explanation);
+    CacheListActivity(4444, R.string.location_permission_request_explanation),
+    CGeoMap(5555, R.string.location_permission_request_explanation),
+    NewMap(6666, R.string.location_permission_request_explanation),
+    EditWaypointActivity(7777, R.string.location_permission_request_explanation),
+    CompassActivity(8888, R.string.location_permission_request_explanation),
+    NavigateAnyPointActivity(9999, R.string.location_permission_request_explanation);
 
     private final int requestCode;
     private final int askAgainResource;

--- a/main/src/cgeo/geocaching/permission/RestartLocationPermissionGrantedCallback.java
+++ b/main/src/cgeo/geocaching/permission/RestartLocationPermissionGrantedCallback.java
@@ -1,0 +1,28 @@
+package cgeo.geocaching.permission;
+
+import cgeo.geocaching.sensors.Sensors;
+import cgeo.geocaching.settings.Settings;
+import cgeo.geocaching.utils.Log;
+
+/**
+ * Extension to @see PermissionGrantedCallback that (in case of granted location permission) initializes the location observables
+ * and then calls executeAfter
+ */
+public abstract class RestartLocationPermissionGrantedCallback extends PermissionGrantedCallback {
+
+    protected RestartLocationPermissionGrantedCallback(final PermissionRequestContext request) {
+        super(request);
+    }
+
+    protected final void execute() {
+        Log.d("RestartLocationPermissionGrantedCallback.execute for " + getRequestCode());
+        final Sensors sensors = Sensors.getInstance();
+        sensors.setupGeoDataObservables(Settings.useGooglePlayServices(), Settings.useLowPowerMode());
+        sensors.setupDirectionObservable();
+
+        executeAfter();
+    }
+
+    protected abstract void executeAfter();
+
+}

